src/utils/nameUtils.ts

```diff
@@ -4,22 +4,117 @@ import { Operation, OperationGroup } from "@azure-tools/codemodel";
 import { getLanguageMetadata } from "./languageHelpers";
 import { TypeDetails, PropertyKind } from "../models/modelDetails";
 
-const ReservedModelNames = ["Error", "Date"];
-
-export enum CasingConvention {
-  Pascal,
-  Camel
+interface ReservedName {
+  name: string;
+  reservedFor: NameType[];
 }
 
 export enum NameType {
   Class,
   File,
   Interface,
-  Property
+  Property,
+  Parameter,
+  Operation,
+  OperationGroup
+}
+
+const Newable = [NameType.Class, NameType.Interface, NameType.OperationGroup];
+
+const ReservedModelNames: ReservedName[] = [
```

Comment:
I like this!  Seems like all of them need `NameType.Parameter` so it feels like that should be checked by default in `guardReservedNames`, but I suppose it makes things less clear to not have it explicitly in the list.

Reply:
Yeah, I thought about that. However by having it there we don't need any NameType specific checks in guardReservedNames and as you said makes it makes it clear and gives us flexibility in case we decide we don't want to suffix a specific reserved word. For instance string would be a valid parameter name